.NET MAUI Community Toolkit - RatingView (#472)
MicrosoftDocs/CommunityToolkit: docs/maui/views/RatingView.md

	xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
	xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
	xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit">
	<VerticalStackLayout>
		<toolkit:RatingView
			EmptyColor="White"
			FilledColor="Blue"
			IsReadOnly="False"
			ItemPadding="3,7,7,3"
			ItemShapeSize="37"
			MaximumRating="5"
			RatingFill="Shape"
			Rating="4.5"
			ItemShape="Star"
			ShapeBorderColor="Red"
			ShapeBorderThickness="1"
			Spacing="3" />
	</VerticalStackLayout>
</ContentPage>
```

The equivalent C# code is:

```csharp
using CommunityToolkit.Maui.Views;

partial class MyPage : ContentPage
{
	public MyPage()
	{
		RatingView ratingView = new()
		{
			EmptyColor = Colors.White,
			FilledColor = Colors.Blue,
			IsReadOnly = false,
			ItemPadding = new Thickness(3,7,7,3),
			ItemShapeSize = 37,
			MaximumRating = 5,
			RatingFill = RatingFillElement.Shape,
			Rating = 4.5,
			ItemShape = RatingViewShape.Star,
			ShapeBorderColor = Colors.Red,
			ShapeBorderThickness = 1,
			Spacing = 3,
		};

		Content = ratingView;
	}
}
```

## Properties

| Property | Type | Description |
|---|---|---|

> | BackgroundColor | `color` | Gets or sets the Color which will fill the background of an element. This is a bindable property. |

Reviewer: I don't think we need BackgroundColor or BindingContext here

Author: You are right.  I'll remove these.

| RatingChanged | `EventHandler<RatingChangedEventArgs>` | Event occurs when the rating is changed. |
| RatingFill | `RatingFillElement` | Gets or sets a value indicating how the fill is applied against the entire rating item or just the shape. The property is of type [`RatingFillElement`](#set-rating-fill) and is an enumeration. The default value of this property is Shape.  This is a bindable property. |
| Rating | `double` | Gets or sets a value indicating the current rating value, allowing for both pre-defined ratings (e.g., from previous user input or stored data) and updates during runtime as the user interacts with the control.  The default value is 0.  This is a bindable property. |
| ShapeBorderColor | `Color` | Gets or sets the border color of the rating item shape. The default value of this is Grey.  This is a bindable property. |
| ShapeBorderThickness | `Thickness` | Gets or sets the border thickness of the rating item shape.  The default value is a Thickness with all values set to 1.  This is a bindable property. |

> | Spacing | `double` | Gets or sets the spacing between rating item elements.  The default value is 10.  This is a bindable property. |

Reviewer: If we are inheriting from HorizontalStackLayout as the next section suggests then we don't need to include Spacing here

The following examples set the is read only property:

```xaml
<toolkit:RatingView
	IsReadOnly="True" />
```

The equivalent C# code is:

```csharp
RatingView readOnlyRatingView = new()
{
	IsReadOnly = True,
};
```

## Set item padding
The `ItemPadding` property is a `Thickness` for the padding between the rating item and its corresponding shape, allowing for finer control over the appearance and layout of the rating items.

The following examples set the item padding property:

```xaml
<toolkit:RatingView
	ItemPadding="3, 7, 7, 3" />
```

The equivalent C# code is:

```csharp
RatingView ratingView = new()
{
	ItemPadding = new Tickness(3, 7, 7, 3),
};
```

## Set item shape size
The `ItemShapeSize` property is a `double` that customizes the shape size to fit the overall design of the application, providing the flexibility to adjust the control to various UI layouts.

The following examples set the item padding property:

```xaml
<toolkit:RatingView
	ItemShapeSize="37" />
```

The equivalent C# code is:

```csharp
RatingView ratingView = new()
{
	ItemShapeSize = 37,
};
```

## Set maximum rating

> The `MaximumRating` property is a `byte` for setting the total number of items (e.g., stars, hearts, etc., or custom shapes) available for rating. This allows for ratings of any scale, such as a 5-star or 10-star system, depending on the needs of the application.

Reviewer: I think this maximum property has a maximum? Should we add that detail here?

---
title: RatingView - .NET MAUI Community Toolkit
author: GeorgeLeithead

> description: "The `RatingView` is designed to provide developers with a flexible and customizable rating mechanism, similar to those used on popular review and feedback platforms."

Reviewer: This is just metadata so we want to remove any formatting.

```suggestion
description: "The RatingView is designed to provide developers with a flexible and customizable rating mechanism, similar to those used on popular review and feedback platforms."
```

ms.date: 09/19/2024
---

# RatingView

> The CommunityToolKit MAUI `RatingView` is a control designed to provide developers with a flexible and customizable rating mechanism, similar to those used on popular review and feedback platforms.

Reviewer: Opt for the official name

```suggestion
The .NET MAUI Community Toolkit `RatingView` is a control designed to provide developers with a flexible and customizable rating mechanism, similar to those used on popular review and feedback platforms.
```

```xaml
<ContentPage
	x:Class="CommunityToolkit.Maui.Sample.Pages.MyPage"
	xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
	xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
	xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit">
	<VerticalStackLayout>
		<toolkit:RatingView
			EmptyColor="White"
			FilledColor="Blue"
			IsReadOnly="False"
			ItemPadding="3,7,7,3"
			ItemShapeSize="37"
			MaximumRating="5"
			RatingFill="Shape"
			Rating="4.5"
			Shape="Star"
			ShapeBorderColor="Red"
			ShapeBorderThickness="1"
			Spacing="3" />
	</VerticalStackLayout>
</ContentPage>
```

The equivalent C# code is:

```csharp
using CommunityToolkit.Maui.Views;

partial class MyPage : ContentPage
{
	public MyPage()
	{
		RatingView ratingView = new()
		{
			EmptyColor = Colors.White,
			FilledColor = Colors.Blue,
			IsReadOnly = false,
			ItemPadding = new Thickness(3,7,7,3),
			ItemShapeSize = 37,
			MaximumRating = 5,
			RatingFill = RatingFillElement.Shape,
			Rating = 4.5,
			Shape = RatingViewShape.Star,
			ShapeBorderColor = Colors.Red,
			ShapeBorderThickness = 1,
			Spacing = 3,
		};

		Content = ratingView;
	}
}
```

## Properties

> | Property | Type | Description |

Reviewer: I think we should only include the properties that we are providing and not those provided by base classes coming out of .NET MAUI. We can link to any MAUI documentation to cover those properties if you want though

Author: Removed base class properties and provided appropriate link.

| Visual | `IVisual` | Gets or sets a IVisual implementation that overrides the visual appearance of an element. This is a bindable property. |
| Width | `double` | Gets the current width of this element. This is a read-only bindable property. |
| WidthRequest | `double` | Gets or sets the desired width override of this element. This is a bindable property. |
| Window | `Window` | Gets the Window that is associated with an element. This is a read-only bindable property. |
| X | `double` | Gets the current X position of this element. This is a read-only bindable property. |
| Y | `double` | Gets the current Y position of this element. This is a read-only bindable property. |
| ZIndex | `int` | Gets or sets the front-to-back z-index of an element within a layout. This is a bindable property. |

## Set custom shape
The `CustomShape` property is a `string` that allows for the defining of custom rating item shape `path`. This feature empowers developers to implement unique designs, such as distinctive symbols, as rating items.

> Only when the `Shape` property is set to `Custom`, will the custom shape path be used.

Reviewer: This feels important

```suggestion
> [!IMPORTANT]
> Only when the `Shape` property is set to `Custom`, will the custom shape path be used.
```

The following examples sets the custom and shape properties:

```xaml
<toolkit:RatingView

> 	CustomShape="M 12 0C5.388 0 0 5.388 0 12s5.388 12 12 12 12-5.38 12-12c0-6.612-5.38-12-12-12z"

Reviewer: What shape does this render? Do we have a screenshot of this?

Author: I didn't want to flood with loads of screen shots, as there is LOADS in the Samples app.  However, in this instance, you are right, and I'll get a screen shot.

	Shape="Custom" />
```

The equivalent C# code is:

```csharp
RatingView ratingView = new()
{
	CustomShape = "M 12 0C5.388 0 0 5.388 0 12s5.388 12 12 12 12-5.38 12-12c0-6.612-5.38-12-12-12z",
	Shape = RatingViewShape.Custom,
};
```

For more information about custom shapes, see [Shapes.Path](/dotnet/api/microsoft.maui.controls.shapes.path).

## Set empty color

> The `EmptyColor` property is a `color` that for the unfilled (empty) rating items. This allows for clear visual differentiation between rated and unrated items.

Reviewer: Typo

```suggestion
The `EmptyColor` property is a `Color` that for the unfilled (empty) rating items. This allows for clear visual differentiation between rated and unrated items.
```

The following examples set the empty color property:

```xaml
<toolkit:RatingView
	EmptyColor="Grey" />
```

The equivalent C# code is:

```csharp
RatingView ratingView = new()
{
	EmptyColor = Colors.Grey,
};
```

## Set filled (rated) color

> The `FilleColor` property is a `color` that will be applied to the filled (rated) portion of each item, offering flexibility in defining the visual aesthetic of the rating items when selected by the user.

Reviewer: Typo

```suggestion
The `FilledColor` property is a `Color` that will be applied to the filled (rated) portion of each item, offering flexibility in defining the visual aesthetic of the rating items when selected by the user.
```

The following examples set the filled color property:

```xaml
<toolkit:RatingView
	FilledColor="Green" />
```

The equivalent C# code is:

```csharp
RatingView ratingView = new()
{
	FilledColor = Colors.Green,
};
```

## Set is read only
The `IsReadOnly` property is a `bool` that will set the control user interface to read only.

The following examples set the is read only property:

```xaml
<toolkit:RatingView
	IsReadOnly="False" />
<toolkit:RatingView
	IsReadOnly="True" />
```

The equivalent C# code is:

```csharp
RatingView editableRatingView = new()
{
	IsReadOnly = False,
};
RatingView readOnlyRatingView = new()
{
	IsReadOnly = True,
};

> ```

Reviewer: Perhaps we just show how to make the control read-only and mention that this defaults to false?

{
	ItemPadding = new Tickness(3, 7, 7, 3),
};
```

## Set item shape size
The `ItemShapeSize` property is a `double` that customizes the shape size to fit the overall design of the application, providing the flexibility to adjust the control to various UI layouts.

The following examples set the item padding property:

```xaml
<toolkit:RatingView
	ItemShapeSize="37" />
```

The equivalent C# code is:

```csharp
RatingView ratingView = new()
{
	ItemShapeSize = 37,
};
```

## Set maximum rating
The `MaximumRating` property is a `byte` for setting the total number of items (e.g., stars, hearts, etc., or custom shapes) available for rating. This allows for ratings of any scale, such as a 5-star or 10-star system, depending on the needs of the application.

If the value is set to 1, the control will toggle the rating between 0 and 1 when clicked/tapped.  If the value is set below the current `Rating`, the rating is adjusted accordingly.

The following examples set the maximum rating property:

```xaml
<toolkit:RatingView
	MaximumRating="7" />
```

The equivalent C# code is:

```csharp
RatingView ratingView = new()
{
	MaximumRating = 7,
};
```

## Set rating fill

> The `RatingFill` property is an `enum` of type `RatingFillElement` for setting how the fill is applied against the entire rating item or just the shape, enabling more nuanced visual presentation, such as filling only the interior of stars or the full item.  The available options are:

Reviewer: I think a screenshot would really help to understand what this does

Author: I didn't want to flood with loads of screen shots, as there is LOADS in the Samples app. However, in this instance, you are right, and I'll get a screen shot.

The `RatingChangedEventArgs` exposes a single property:
- `Rating` - This is the new rating value.

The following examples show how to attach the  event:

```xaml
<toolkit:RatingView
	RatingChanged="RatingView_RatingChanged" />
```

The equivalent C# code is:

```csharp
	RatingView ratingView = new();
	ratingView.RatingChanged += RatingView_RatingChanged;
```

The following example is the code behind to handle the event:

```csharp
void RatingView_RatingChanged(object sender, RatingChangedEventArgs e)
{
	double newRating = e.Rating;
	// The developer can then perform further actions (such as save to DB).
}
```

## Set shape
The `Shape` property is an `enum` of type `RatingViewShape` for setting the rating item shape of the rating items, such as stars, circles, like, dislike, or any other commonly used rating icons..  The available options are:

- `Star` - (default)
- `Heart`
- `Circle`
- `Like`
- `Dislike`
- `Custom` - If set and `CustomShape` is NULL or empty, defaults to `Star`

The following examples set the rating fill property:

> ```xaml

Reviewer: Perhaps some screenshots of this would be helpful

<toolkit:RatingView
	ShapeBorderColor="Grey" />
```

The equivalent C# code is:

```csharp
RatingView ratingView = new()
{
	ShapeBorderColor = Colors.Grey,
};
```

## Set shape border thickness
The `ShapeBorderThickness` is a `double` for setting the thickness of the shape border. This provides additional flexibility to create visually distinct and stylized rating shapes with custom borders.

The following examples set the shape border thickness property:

```xaml
<toolkit:RatingView
	ShapeBorderThickness="3" />
```

The equivalent C# code is:

```csharp
RatingView ratingView = new()
{
	ShapeBorderThickness = 3,
};
```

## Set spacing
The `Spacing` is a `double` for setting the spacing between each rating item.

The following examples set the shape border thickness property:

```xaml
<toolkit:RatingView
	Spacing="7" />
```

The equivalent C# code is:

```csharp
RatingView ratingView = new()
{
	Spacing = 7,
};
```

## Examples

You can find examples of this control in action in the .NET MAUI Community Toolkit Sample Application:
- [XAML Syntax](https://github.com/CommunityToolkit/Maui/blob/main/samples/CommunityToolkit.Maui.Sample/Pages/Views/RatingView/RatingViewXamlPage.xaml)

> - [C# Syntax](https://github.com/CommunityToolkit/Maui/blob/main/samples/CommunityToolkit.Maui.Sample/Pages/Views/RatingView/RatingViewCsharpPage.xaml)

Reviewer: It's possibly not important but I find it intriguing that the Csharp page example is a xaml file, is this correct?

Author: Copy paste issue here, where I failed to update the extension.  Will fix.
